### PR TITLE
pipenv 3000 install (behavioral refactor)

### LINKED ISSUE #5992: `pipenv install <package>` upgrades unrelated packages
### Issue description

Installing a new package with `pipenv install <package>` upgrades unrelated packages.

### Expected result

It does not modify unrelated packages and their dependencies.

### Actual result

It modifies unrelated packages and their dependencies.

### Steps to replicate

1. Create a `Pipfile` with a dependency:
   1. whose version in the `Pipfile` is `*` and

      <details>
        <summary>Example</summary>

        ```sh
        cat > Pipfile <<EOF
        [[source]]
        url = "https://pypi.org/simple"
        verify_ssl = true
        name = "pypi"

        [packages]
        idna = "*"

        [dev-packages]

        [requires]
        python_version = "3.11"
        EOF
        ```
      </details>

   2. whose version in the `Pipfile.lock` is not the current version.

      <details>
        <summary>Example</summary>

        ```sh
        cat > Pipfile.lock <<EOF
        {
            "_meta": {
                "hash": {
                    "sha256": "3c314687ccfa77044481a5041894405bf4ff3149d42750829e3e04632cc6ecb9"
                },
                "pipfile-spec": 6,
                "requires": {
                    "python_version": "3.11"
                },
                "sources": [
                    {
                        "name": "pypi",
                        "url": "https://pypi.org/simple",
                        "verify_ssl": true
                    }
                ]
            },
            "default": {
                "idna": {
                    "hashes": [
                        "sha256:b307872f855b18632ce0c21c5e45be78c0ea7ae4c15c828c20788b26921eb3f6",
                        "sha256:b97d804b1e9b523befed77c48dacec60e6dcb0b5391d57af6a65a312a90648c0"
                    ],
                    "index": "pypi",
                    "markers": "python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'",
                    "version": "==2.10"
                }
            },
            "develop": {}
        }
        EOF
        ```
      </details>

2. Install the existing package(s).

   <details>
     <summary>Example</summary>

     ```
     $ python -m pipenv install --verbose
     Using python: 3.11
     Path to python: $HOME/.pyenv/versions/3.11.6/bin/python3
     Creating a virtualenv for this project...
     Pipfile: $HOME/tmp/pipenv-install-upgrade/Pipfile
     Using $HOME/.pyenv/versions/3.11.6/bin/python3 (3.11.6) to create virtualenv...
     ⠙ Creating virtual environment...created virtual environment CPython3.11.6.final.0-64 in 153ms
       creator CPython3Posix(dest=$HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq, clear=False, no_vcs_ignore=False, global=False)
       seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=$HOME/Library/Application Support/virtualenv)
         added seed packages: pip==23.2.1, setuptools==68.2.2, wheel==0.41.2
       activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

     ✔ Successfully created virtual environment!
     Virtualenv location: $HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq
     Installing dependencies from Pipfile.lock (c6ecb9)...
     Writing supplied requirement line to temporary file: "idna==2.10; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'
     --hash=sha256:b307872f855b18632ce0c21c5e45be78c0ea7ae4c15c828c20788b26921eb3f6 --hash=sha256:b97d804b1e9b523befed77c48dacec60e6dcb0b5391d57af6a65a312a90648c0"
     Install Phase: Standard Requirements
     Preparing Installation of "idna==2.10; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'
     --hash=sha256:b307872f855b18632ce0c21c5e45be78c0ea7ae4c15c828c20788b26921eb3f6 --hash=sha256:b97d804b1e9b523befed77c48dacec60e6dcb0b5391d57af6a65a312a90648c0"
     $ $HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/bin/python
     $HOME/.pyenv/versions/3.11.6/lib/python3.11/site-packages/pipenv/patched/pip/__pip-runner__.py install -i https://pypi.org/simple --no-input
     --upgrade --no-deps -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-6_8mrk_n-requirements/pipenv-qdfcg2qe-hashed-reqs.txt
     Using source directory: '$HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/src'
     Collecting idna==2.10 (from -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-6_8mrk_n-requirements/pipenv-qdfcg2qe-hashed-reqs.txt (line 1))

       Downloading idna-2.10-py2.py3-none-any.whl (58 kB)

          ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 2.7 MB/s eta 0:00:00

     Installing collected packages: idna

     Successfully installed idna-2.10


     To activate this project's virtualenv, run pipenv shell.
     Alternatively, run a command inside the virtualenv with pipenv run.
     ```
   </details>

3. Install a new package with `pipenv install <package>`.

   <details>
     <summary>Example</summary>

     ```
     $ python -m pipenv install leftpad --verbose
     Installing leftpad...
     Resolving leftpad...
     Added leftpad to Pipfile's [packages] ...
     ✔ Installation Succeeded
     Pipfile.lock (c6ecb9) out of date, updating to (fe4271)...
     Locking [packages] dependencies...
     Building requirements...
     Resolving dependencies...
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting()
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('idna'), None)
     INFO:pipenv.patched.pip._internal.operations.prepare:Collecting idna (from -r
     /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-tyevi2hm-requirements/pipenv-qfc2bo33-constraints.txt (line 3))
     INFO:pipenv.patched.pip._internal.network.download:Using cached idna-3.4-py3-none-any.whl (61 kB)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.adding_requirement(SpecifierRequirement('leftpad'), None)
     INFO:pipenv.patched.pip._internal.operations.prepare:Collecting leftpad (from -r
     /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-tyevi2hm-requirements/pipenv-qfc2bo33-constraints.txt (line 2))
     INFO:pipenv.patched.pip._internal.network.download:Using cached leftpad-0.1.2-py3-none-any.whl (1.8 kB)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(0)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.pinning(LinkCandidate('https://files.pythonhosted.org/packages/fc/34/3030de6f1370931
     b9dbb4dad48f6ab1015ab1d32447850b9fc94e60097be/idna-3.4-py3-none-any.whl (from https://pypi.org/simple/idna/) (requires-python:>=3.5)'))
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending_round(0, state)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(1)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.pinning(LinkCandidate('https://files.pythonhosted.org/packages/c2/92/b15f48c2fb7c18f
     f0a21d77c6c1e0f3ae743b8d5b0516841cd290ec5e283/leftpad-0.1.2-py3-none-any.whl (from https://pypi.org/simple/leftpad/)'))
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending_round(1, state)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.starting_round(2)
     INFO:pipenv.patched.pip._internal.resolution.resolvelib.reporter:Reporter.ending(State(mapping=OrderedDict([('idna',
     LinkCandidate('https://files.pythonhosted.org/packages/fc/34/3030de6f1370931b9dbb4dad48f6ab1015ab1d32447850b9fc94e60097be/idna-3.4-py3-none-any.whl (from
     https://pypi.org/simple/idna/) (requires-python:>=3.5)')), ('leftpad',
     LinkCandidate('https://files.pythonhosted.org/packages/c2/92/b15f48c2fb7c18ff0a21d77c6c1e0f3ae743b8d5b0516841cd290ec5e283/leftpad-0.1.2-py3-none-any.whl (from
     https://pypi.org/simple/leftpad/)'))]), criteria={'idna': Criterion((SpecifierRequirement('idna'), via=None)), 'leftpad':
     Criterion((SpecifierRequirement('leftpad'), via=None))}, backtrack_causes=[]))
     ⠴ Locking...
     ✔ Success!
     Locking [dev-packages] dependencies...
     Updated Pipfile.lock (e7ca18f416e18e6e31b172657ee3d5e41fb06ef4a01cabf2b66311cad5fe4271)!
     Installing dependencies from Pipfile.lock (fe4271)...
     Writing supplied requirement line to temporary file: "idna==3.4; python_version >= '3.5'
     --hash=sha256:814f528e8dead7d329833b91c5faa87d60bf71824cd12a7530b5526063d02cb4 --hash=sha256:90b77e79eaa3eba6de819a0c442c0b4ceefc341a7a2ab77d7562bf49f425c5c2"
     Writing supplied requirement line to temporary file: 'leftpad==0.1.2 --hash=sha256:8ff2df0f8dbe6537d7f5f7cfb579b628ee81d02556971f2bb5992f6dffacc60e
     --hash=sha256:e38a2ebc18e0ce3adb4116708c9d8b0ec1d15414eafa4d236be7db893e8dc911'
     Install Phase: Standard Requirements
     Preparing Installation of "idna==3.4; python_version >= '3.5' --hash=sha256:814f528e8dead7d329833b91c5faa87d60bf71824cd12a7530b5526063d02cb4
     --hash=sha256:90b77e79eaa3eba6de819a0c442c0b4ceefc341a7a2ab77d7562bf49f425c5c2"
     Preparing Installation of 'leftpad==0.1.2 --hash=sha256:8ff2df0f8dbe6537d7f5f7cfb579b628ee81d02556971f2bb5992f6dffacc60e
     --hash=sha256:e38a2ebc18e0ce3adb4116708c9d8b0ec1d15414eafa4d236be7db893e8dc911'
     $ $HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/bin/python
     $HOME/.pyenv/versions/3.11.6/lib/python3.11/site-packages/pipenv/patched/pip/__pip-runner__.py install -i https://pypi.org/simple --no-input
     --upgrade --no-deps -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-i0ab865p-requirements/pipenv-0z3oitef-hashed-reqs.txt
     Using source directory: '$HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/src'
     Collecting idna==3.4 (from -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-i0ab865p-requirements/pipenv-0z3oitef-hashed-reqs.txt (line 1))

       Using cached idna-3.4-py3-none-any.whl (61 kB)

     Collecting leftpad==0.1.2 (from -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-i0ab865p-requirements/pipenv-0z3oitef-hashed-reqs.txt (line 2))

       Using cached leftpad-0.1.2-py3-none-any.whl (1.8 kB)

     Installing collected packages: leftpad, idna

       Attempting uninstall: idna

         Found existing installation: idna 2.10

         Uninstalling idna-2.10:

           Successfully uninstalled idna-2.10

     Successfully installed idna-3.4 leftpad-0.1.2

     Install Phase: Editable Requirements
     Preparing Installation of "idna==3.4; python_version >= '3.5' --hash=sha256:814f528e8dead7d329833b91c5faa87d60bf71824cd12a7530b5526063d02cb4
     --hash=sha256:90b77e79eaa3eba6de819a0c442c0b4ceefc341a7a2ab77d7562bf49f425c5c2"
     Preparing Installation of 'leftpad==0.1.2 --hash=sha256:8ff2df0f8dbe6537d7f5f7cfb579b628ee81d02556971f2bb5992f6dffacc60e
     --hash=sha256:e38a2ebc18e0ce3adb4116708c9d8b0ec1d15414eafa4d236be7db893e8dc911'
     $ $HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/bin/python
     $HOME/.pyenv/versions/3.11.6/lib/python3.11/site-packages/pipenv/patched/pip/__pip-runner__.py install -i https://pypi.org/simple --no-input
     --upgrade --no-deps -r /var/folders/g6/2hbvhk753g96h2yd11y5x6280000gp/T/pipenv-i0ab865p-requirements/pipenv-ix2tblys-reqs.txt
     Using source directory: '$HOME/.local/share/virtualenvs/pipenv-install-upgrade-ar47k8rq/src'


     To activate this project's virtualenv, run pipenv shell.
     Alternatively, run a command inside the virtualenv with pipenv run.
     ```
   </details>

---

<details>
  <summary>$ python -m pipenv --support</summary>

  Pipenv version: `'2023.10.20'`

  Pipenv location: `'$HOME/.pyenv/versions/3.11.6/lib/python3.11/site-packages/pipenv'`

  Python location: `'$HOME/.pyenv/versions/3.11.6/bin/python'`

  OS Name: `'posix'`

  User pip version: `'23.3'`

  user Python installations found:

  PEP 508 Information:

  ```
  {'implementation_name': 'cpython',
   'implementation_version': '3.11.6',
   'os_name': 'posix',
   'platform_machine': 'arm64',
   'platform_python_implementation': 'CPython',
   'platform_release': '23.0.0',
   'platform_system': 'Darwin',
   'platform_version': 'Darwin Kernel Version 23.0.0: Fri Sep 15 14:41:43 PDT '
                       '2023; root:xnu-10002.1.13~1/RELEASE_ARM64_T6000',
   'python_full_version': '3.11.6',
   'python_version': '3.11',
   'sys_platform': 'darwin'}
  ```
</details>

### LINKED ISSUE #6048: "pipenv install" is unclear to me
It is unclear to me what the purpose of `pipenv install` is.

Usually I do install packages via `pip` in a virtual environment. So what is the advantage of `pipenv install`?

### LINKED ISSUE #6070: `pipenv install` should do minimal changes to `Pipfile.lock`
`pipenv install` should do minimal changes to `Pipfile.lock`. Now it updates versions of packages unrelated to a package specified in `pipenv install` wreaking havoc in package versions.

`npm` and `yarn` do the correct thing on this (I assume).

### PR DESCRIPTION
**Pull request by matteius**
### The issue

It has been long requested that:
`pipenv install` have the same behavior as `pipenv sync` when no packages are requested
and `pipenv install package` has the behavior of `pipenv update` 
which means only `pipenv lock` will regenerate the whole lock.

Fixes #5992
Fixes #6070
Fixes #6048

### The checklist

* [X] Associated issue
* [X] A news fragment in the `news/` directory to describe this fix with the extension `.bugfix.rst`, `.feature.rst`, `.behavior.rst`, `.doc.rst`. `.vendor.rst`. or `.trivial.rst` (this will appear in the release changelog). Use semantic line breaks and name the file after the issue number or the PR #.

### CODE REVIEW

**Comment by oz123** [comment]
This PR looks good. What's still missing is an update to docs.

**Review by oz123 — CHANGES_REQUESTED** [review]
We should definitely add documentation about this.

**Comment by Kroppeb** [comment]
Huh, I'm a bit confused. I only see the logic for `pipenv install package` having been changed. I don't see where you changed the behavior of `pipenv install`.

I know some people at my work sometimes edit the pipfile directly and then run `pipenv install`. Maybe it would be beneficial to see if the lockfile is up to date and if not use the old logic of lock & sync?

**Comment by matteius** [comment]
> Huh, I'm a bit confused. I only see the logic for pipenv install package having been changed. 

Thanks for checking, I probably missed something in the refactor -- will revisit soon.

**Comment by matteius** [comment]
@Kroppeb I believe the PR is now in a good spot and does what I expect it to.

@oz123 I made the corresponding documentation updates as well, please re-review.

File: docs/pipfile.md [hunk 1]
```diff
@@ -267,7 +267,7 @@ All named categories (other than the special default/develop) will use the categ
 ## General Notes and Recommendations
 
 - Keep both `Pipfile` and `Pipfile.lock` in version control.
-- `pipenv install` adds specifiers to `Pipfile` and rebuilds the lock file based on the Pipfile specs, by utilizing the internal resolver of `pip`.
+- `pipenv install pacckage-name` adds specifiers to `Pipfile` and rebuilds the lock file based on the Pipfile specs, by utilizing the internal resolver of `pip`.
```
Comment (sanspareilsmyn) [inline docs/pipfile.md:270]: @matteius I think there's a typo here.

File: docs/commands.md [hunk 2]
```diff
@@ -5,13 +5,13 @@ The commands reference for pipenv (incomplete)
 ## install
 
 ``$ pipenv install`` is used for installing packages into the pipenv virtual environment
-and updating your Pipfile and Pipfile.lock.
+and updating your Pipfile and Pipfile.lock in the case of adding new packages.
```
Comment (sanspareilsmyn) [inline docs/commands.md:8]: 👍

File: pipenv/routines/update.py [hunk 3]
```diff
@@ -42,6 +42,10 @@ def do_update(
         site_packages=site_packages,
         clear=clear,
     )
+    if packages is None:
```
Comment (sanspareilsmyn) [inline pipenv/routines/update.py:45]: In my opinion, this would be more pythonic if we go like this :)

```python
packages = [p for p in (packages or []) if p]
editable = [p for p in (editable_packages or []) if p]
```

Comment (matteius) [reply pipenv/routines/update.py:45]: Thanks!

**Review by sanspareilsmyn — APPROVED** [review]

File: docs/pipfile.md [hunk 4]
```diff
@@ -240,7 +240,7 @@ where the lock file was last updated, even on untrusted networks.
 We recommend designing CI/CD deployments whereby the build does not alter the lock file as a side effect.
 In other words, you can use `pipenv lock` or `pipenv upgrade` to adjust your lockfile through local development,
 the PR process and approve those lock changes before deploying to production that version of the lockfile.
-In other words avoid having your CI issue `lock`, `update`, `upgrade` `uninstall` or `install` commands that will relock.
+In other words avoid having your CI issue `lock`, `update`, `upgrade` `uninstall` commands that will relock.
 Note:  It is counterintuitive that `pipenv install` re-locks and `pipenv sync` or `pipenv install --deploy` does not.
```
Comment (oz123) [inline docs/pipfile.md:244]: This seems contrary to other changes in this PR.

> It is counterintuitive that `pipenv install` re-locks and `pipenv sync` or `pipenv install --deploy` does not.

This should be updated accordingly.

**Review by oz123 — CHANGES_REQUESTED** [review]

File: docs/pipfile.md (hunk 4 above)
Comment (matteius) [reply docs/pipfile.md]: Good find; All set!

**Review by Kroppeb — COMMENTED** [review]
Is there still a way to lock and install in 1 command. If not, maybe adding a `pipenv install --relock` could be interesting? Could be done in a later pr though I suppose.

File: docs/advanced.md [hunk 5]
```diff
@@ -333,12 +333,8 @@ Here's an example `tox.ini` for both local and external testing:
...
 
-You might also want to add `--ignore-pipfile` to `pipenv install`, as to not accidentally modify the lock-file on each test run. This causes Pipenv
-to ignore changes to the `Pipfile` and (more importantly) prevents it from adding the current environment to `Pipfile.lock`. This might be important as the current environment (i.e. the virtualenv provisioned by tox) will usually
-contain the current project (which may or may not be desired) and additional dependencies from `tox`'s `deps` directive. The initial provisioning may
-alternatively be disabled by adding `skip_install = True` to tox.ini.
-
-This method requires you to be explicit about updating the lock-file, which is probably a good idea in any case.
+Some legacy users supply `--ignore-pipfile` to `pipenv install`. This causes Pipenv to ignore the contents of the `Pipfile.lock`.
+This method is dangerous as you do not get the guarantees of repeatable package installs and package hash verifications that are provided by utilizing the `Pipfile.lock`.
```
[the remark this answers is not in the document]
Comment (matteius) [reply docs/advanced.md:337]: No, you are right, it really doesn't do anything now I guess.   I'll adjust the docs further.

**Review by oz123 — APPROVED** [review]
